src/theory/bv/bvgauss.cpp

```diff
@@ -0,0 +1,692 @@
+ ** Top contributors (to current version):
+ **   Aina Niemetz
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2017 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Gaussian Elimination preprocessing pass.
+ **
+ ** Simplify a given equation system modulo a (prime) number via Gaussian
+ ** Elimination if possible.
+ **/
+
+#include "theory/bv/bvgauss.h"
+
+#include <iostream>
+#include <stack>
+#include <unordered_map>
+
+#include "theory/bv/theory_bv_utils.h"
+
+using namespace CVC4;
+using namespace std;
+
+namespace CVC4 {
+namespace theory {
+namespace bv {
+
+unsigned BVGaussElim::getMinBwExpr(Node expr)
+{
+  stack<Node> visit;
+  unordered_map<Node, unsigned, NodeHashFunction> visited;
+  unordered_map<Node, unsigned, NodeHashFunction>::iterator it;
+
+  visit.push(expr);
+  while (!visit.empty())
+  {
+    Node n = visit.top();
+    visit.pop();
+    it = visited.find(n);
+    if (it == visited.end())
+    {
+      visited[n] = 0;
+      visit.push(n);
+      for (Node child : n) visit.push(child);
+    }
+    else if (it->second == 0)
+    {
+      Kind k = n.getKind();
+      switch (k)
+      {
+        case kind::CONST_BITVECTOR:
+        {
+          unsigned bwval;
+          Integer val = n.getConst<BitVector>().getValue();
+          for (bwval = 0; val > 0; val = val.divByPow2(1), ++bwval)
+            ;
+          visited[n] = bwval;
+          break;
+        }
+
+        case kind::BITVECTOR_EXTRACT:
+        {
+          unsigned w = utils::getSize(n);
+          visited[n] = w >= visited[n[0]] ? visited[n[0]] : w;
+          Assert(visited[n] <= visited[n[0]]);
+          break;
+        }
+
+        case kind::BITVECTOR_ZERO_EXTEND:
+        {
+          visited[n] = visited[n[0]];
+          break;
+        }
+
+        case kind::BITVECTOR_MULT:
+        {
+          unsigned w = 0;
+          for (unsigned i = 0, nc = n.getNumChildren(); i < nc; ++i)
+            w += visited[n[i]];
+          if (w > utils::getSize(n)) return 0; /* overflow */
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_CONCAT:
+        {
+          unsigned w;
+          Node zero = utils::mkZero(utils::getSize(n[0]));
+          if (n[0] == zero)
+          {
+            w = visited[n[1]]
+                /* add bw of children n[2] to n[n.getNumChildren()-1] */
+                + utils::getSize(n) - utils::getSize(n[0])
+                - utils::getSize(n[1]);
+          }
+          else
+          {
+            w = visited[n[0]]
+              /* add bw of children n[1] to n[n.getNumChildren()-1] */
+              +  utils::getSize(n) - utils::getSize(n[0]);
+          }
+          visited[n] = w;
+          break;
+        }
+
+        case kind::BITVECTOR_UDIV_TOTAL:
```

Comment:
Actually, I just realized that for udiv it is not even possible to minimize the bit-width. Consider the worst case: division by 0 (since x / 0 = -1).

Reply:
OK, I've been thinking this through again and came to the conclusion that division by zero is irrelevant for determining the minimum bit-width since unsigned division can not overflow. Hence, an expression like the following would have a sufficiently large bit-width:

Assume that `x`, `y` and `z` are bit-vector variables of bit-width 2:
```
(bvadd (bvudiv (concat (_ bv0 2) x) (concat (_ bv0 2) y)) (concat (_ bv0 2) z))
```

I agree with @4tXJ7f  that only the size of the dividend is relevant for unsigned division.